sstable: index out of range in decodeZstd

https://github.com/cockroachdb/pebble/actions/runs/10726746532/job/29747482892?pr=3917

```
--- FAIL: TestCompressionRoundtrip (0.01s)
    compression_test.go:22: seed 1725564158549686258
    --- FAIL: TestCompressionRoundtrip/ZSTD (0.00s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

goroutine 12 [running]:
testing.tRunner.func1.2({0xaec880, 0xc000031c80})
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1634 +0x6b6
panic({0xaec880?, 0xc000031c80?})
	/opt/hostedtoolcache/go/1.22.6/x64/src/runtime/panic.go:7[70](https://github.com/cockroachdb/pebble/actions/runs/10726746532/job/29747482892?pr=3917#step:4:71) +0x132
github.com/DataDog/zstd.DecompressInto.func1(0xc00006bd20, 0xc00006bd38)
	/home/runner/go/pkg/mod/github.com/!data!dog/zstd@v1.5.6-0.20230824185856-869dae002e5e/zstd.go:169 +0x1b8
github.com/DataDog/zstd.DecompressInto({0x0, 0x0, 0x0}, {0xc00018e221, 0x9, 0x1e})
	/home/runner/go/pkg/mod/github.com/!data!dog/zstd@v1.5.6-0.20230824185856-869dae002e5e/zstd.go:1[72](https://github.com/cockroachdb/pebble/actions/runs/10726746532/job/29747482892?pr=3917#step:4:73) +0x49
github.com/cockroachdb/pebble/sstable/block.decodeZstd(...)
	/home/runner/work/pebble/pebble/sstable/block/compression_cgo.go:30
github.com/cockroachdb/pebble/sstable/block.DecompressInto(0x7, {0xc00018e221, 0x9, 0x1e}, {0x0, 0x0, 0x0})
	/home/runner/work/pebble/pebble/sstable/block/compression.go:148 +0xa9
github.com/cockroachdb/pebble/sstable/block.decompress(0x7, {0xc00018e220, 0xa, 0x1f})
	/home/runner/work/pebble/pebble/sstable/block/compression_test.go:82 +0x111
github.com/cockroachdb/pebble/sstable/block.TestCompressionRoundtrip.func1(0xc0001c2820)
	/home/runner/work/pebble/pebble/sstable/block/compression_test.go:34 +0xd8
testing.tRunner(0xc0001c2820, 0xc000013920)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 9
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1742 +0x826

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0001c2000, {0xb2a2[73](https://github.com/cockroachdb/pebble/actions/runs/10726746532/job/29747482892?pr=3917#step:4:74), 0x18}, 0xb5b478)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1750 +0x851
testing.runTests.func1(0xc0001c2000)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:2161 +0x86
testing.tRunner(0xc0001c2000, 0xc00003fb10)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1689 +0x21f
testing.runTests(0xc000013788, {0xe9e9c0, 0x4, 0x4}, {0xc00003fbb8?, 0xc00003fc00?, 0xef82e0?})
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:2159 +0x8bf
testing.(*M).Run(0xc00016bc20)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:2027 +0xf18
main.main()
	_testmain.go:53 +0x2be

goroutine 9 [chan receive]:
testing.(*T).Run(0xc0001c2340, {0xb20f36, 0x4}, 0xc000013920)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1750 +0x851
github.com/cockroachdb/pebble/sstable/block.TestCompressionRoundtrip(0xc0001c2340)
	/home/runner/work/pebble/pebble/sstable/block/compression_test.go:26 +0x385
testing.tRunner(0xc0001c2340, 0xb5b478)
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 1
	/opt/hostedtoolcache/go/1.22.6/x64/src/testing/testing.go:1[74](https://github.com/cockroachdb/pebble/actions/runs/10726746532/job/29747482892?pr=3917#step:4:75)2 +0x826
FAIL	github.com/cockroachdb/pebble/sstable/block	0.0[76](https://github.com/cockroachdb/pebble/actions/runs/10726746532/job/29747482892?pr=3917#step:4:77)s
```

--- a/sstable/block/compression_cgo.go
+++ b/sstable/block/compression_cgo.go
@@ -11,6 +11,7 @@ import (
 	"bytes"
 
 	"github.com/DataDog/zstd"
+	"github.com/cockroachdb/errors"
 )
 
 // UseStandardZstdLib indicates whether the zstd implementation is a port of the
@@ -27,6 +28,12 @@ const UseStandardZstdLib = true
 // decodeZstd decompresses src with the Zstandard algorithm. The destination
 // buffer must already be sufficiently sized, otherwise decodeZstd may error.
 func decodeZstd(dst, src []byte) ([]byte, error) {
+	if len(src) == 0 {
+		return nil, errors.Errorf("decodeZstd: empty src buffer")
+	}
+	if len(dst) == 0 {
+		return nil, errors.Errorf("decodeZstd: empty dst buffer")
+	}
 	n, err := zstd.DecompressInto(dst, src)
 	// NB: zstd.DecompressInto may return n < 0 if err != nil.
 	if err != nil {

--- a/sstable/block/compression_test.go
+++ b/sstable/block/compression_test.go
@@ -24,11 +24,11 @@ func TestCompressionRoundtrip(t *testing.T) {
 
 	for compression := DefaultCompression + 1; compression < NCompression; compression++ {
 		t.Run(compression.String(), func(t *testing.T) {
-			payload := make([]byte, rng.Intn(10<<10 /* 10 KiB */))
+			payload := make([]byte, 1+rng.Intn(10<<10 /* 10 KiB */))
 			rng.Read(payload)
 			// Create a randomly-sized buffer to house the compressed output. If it's
 			// not sufficient, Compress should allocate one that is.
-			compressedBuf := make([]byte, rng.Intn(1<<10 /* 1 KiB */))
+			compressedBuf := make([]byte, 1+rng.Intn(1<<10 /* 1 KiB */))
 
 			btyp, compressed := compress(compression, payload, compressedBuf)
 			v, err := decompress(btyp, compressed)

